File: examples/websocket_client.py
"""
Example WebSocket collaboration client for Giga-PDF.

This example demonstrates how to connect to the Giga-PDF collaboration
server and interact with other users in real-time.

Usage:
    python examples/websocket_client.py --document-id <uuid> --token <jwt>
"""

import argparse
import asyncio
import logging
from datetime import datetime

import socketio

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)


class CollaborationClient:
    """
    WebSocket collaboration client for Giga-PDF.

    Handles real-time collaboration including user presence,
    element locking, cursor tracking, and document updates.
    """

    def __init__(self, server_url: str, document_id: str, auth_token: str):
        """
        Initialize collaboration client.

        Args:
            server_url: WebSocket server URL (e.g., http://localhost:8000/ws).
            document_id: Document UUID to collaborate on.
            auth_token: JWT authentication token.
        """
        self.server_url = server_url
        self.document_id = document_id
        self.auth_token = auth_token
        self.sio = socketio.AsyncClient(logger=True, engineio_logger=True)
        self.session_id = None
        self.user_color = None
        self.active_users = {}
        self.locked_elements = set()

        self.setup_event_handlers()

    def setup_event_handlers(self):
        """Set up Socket.IO event handlers."""

        @self.sio.on("connect")
        async def on_connect():
            """Handle connection to server."""
            logger.info("Connected to collaboration server")
            await self.join_document()

        @self.sio.on("disconnect")
        async def on_disconnect():
            """Handle disconnection from server."""
            logger.info("Disconnected from collaboration server")
            self.active_users.clear()
            self.locked_elements.clear()

        @self.sio.on("user:joined")
        async def on_user_joined(data):
            """Handle user joining the document."""
            user_id = data["user_id"]
            user_name = data["user_name"]
            user_color = data["user_color"]

            self.active_users[user_id] = {
                "name": user_name,
                "color": user_color,
            }

            logger.info(
                f"User joined: {user_name} ({user_id}) with color {user_color}"
            )

        @self.sio.on("user:left")
        async def on_user_left(data):
            """Handle user leaving the document."""
            user_id = data["user_id"]
            user_name = data["user_name"]

            if user_id in self.active_users:
                del self.active_users[user_id]

            logger.info(f"User left: {user_name} ({user_id})")

        @self.sio.on("element:locked")
        async def on_element_locked(data):
            """Handle element being locked by another user."""
            element_id = data["element_id"]
            locked_by = data["locked_by_user_name"]
            expires_at = data["expires_at"]

            self.locked_elements.add(element_id)

            logger.info(
                f"Element {element_id} locked by {locked_by} until {expires_at}"
            )

        @self.sio.on("element:unlocked")
        async def on_element_unlocked(data):
            """Handle element being unlocked."""
            element_id = data["element_id"]

            self.locked_elements.discard(element_id)

            logger.info(f"Element {element_id} unlocked")

        @self.sio.on("cursor:moved")
        async def on_cursor_moved(data):
            """Handle cursor movement from another user."""
            user_id = data["user_id"]
            user_name = data["user_name"]
            page = data["page"]
            x = data["x"]
            y = data["y"]

            logger.debug(
                f"Cursor from {user_name}: page {page}, position ({x:.1f}, {y:.1f})"
            )

        @self.sio.on("document:updated")
        async def on_document_updated(data):
            """Handle document update from another user."""
            user_name = data["user_name"]
            update_type = data["update_type"]
            affected_elements = data.get("affected_elements", [])
            affected_pages = data.get("affected_pages", [])

            logger.info(
                f"Document updated by {user_name}: {update_type} "
                f"(elements: {len(affected_elements)}, pages: {affected_pages})"
            )

    async def connect(self):
        """Connect to the WebSocket server."""
        logger.info(f"Connecting to {self.server_url}...")

        await self.sio.connect(
            self.server_url,
            auth={"token": self.auth_token},
            transports=["websocket", "polling"],
        )

    async def join_document(self):
        """Join the document collaboration room."""
        logger.info(f"Joining document {self.document_id}...")

        response = await self.sio.call(
            "join_document",
            {"document_id": self.document_id},
            timeout=10,
        )

        if response.get("success"):
            self.session_id = response["session"]["session_id"]
            self.user_color = response["session"]["user_color"]

            logger.info(f"Joined document successfully!")
            logger.info(f"Session ID: {self.session_id}")
            logger.info(f"Your color: {self.user_color}")

            # Log active users
            active_users = response.get("active_users", [])
            if active_users:
                logger.info(f"Active users ({len(active_users)}):")
                for user in active_users:
                    logger.info(
                        f"  - {user['user_name']} ({user['user_id']}) "
                        f"with color {user['user_color']}"
                    )
                    self.active_users[user["user_id"]] = {
                        "name": user["user_name"],
                        "color": user["user_color"],
                    }
            else:
                logger.info("You are the first user in this document")

            # Log active locks
            active_locks = response.get("active_locks", [])
            if active_locks:
                logger.info(f"Active locks ({len(active_locks)}):")
                for lock in active_locks:
                    logger.info(
                        f"  - Element {lock['element_id']} "
                        f"locked by {lock['locked_by_user_id']}"
                    )
                    self.locked_elements.add(lock["element_id"])
        else:
            error = response.get("error", "Unknown error")
            logger.error(f"Failed to join document: {error}")

    async def leave_document(self):
        """Leave the document collaboration room."""
        logger.info("Leaving document...")

        response = await self.sio.call("leave_document", {}, timeout=5)

        if response.get("success"):
            logger.info("Left document successfully")
            self.active_users.clear()
            self.locked_elements.clear()
        else:
            error = response.get("error", "Unknown error")
            logger.error(f"Failed to leave document: {error}")

    async def lock_element(self, element_id: str) -> bool:
        """
        Lock an element for editing.

        Args:
            element_id: Element UUID to lock.

        Returns:
            bool: True if lock acquired, False otherwise.
        """
        logger.info(f"Locking element {element_id}...")

        response = await self.sio.call(
            "element:lock",
            {"element_id": element_id},
            timeout=5,
        )

        if response.get("success"):
            logger.info(
                f"Element locked until {response['lock']['expires_at']}"
            )
            self.locked_elements.add(element_id)
            return True
        else:
            error = response.get("error", "Unknown error")
            locked_by = response.get("locked_by")
            expires_at = response.get("expires_at")

            logger.warning(
                f"Failed to lock element: {error} "
                f"(locked by {locked_by} until {expires_at})"
            )
            return False

    async def unlock_element(self, element_id: str):
        """
        Unlock an element.

        Args:
            element_id: Element UUID to unlock.
        """
        logger.info(f"Unlocking element {element_id}...")

        response = await self.sio.call(
            "element:unlock",
            {"element_id": element_id},
            timeout=5,
        )

        if response.get("success"):
            logger.info("Element unlocked successfully")
            self.locked_elements.discard(element_id)
        else:
            error = response.get("error", "Unknown error")
            logger.warning(f"Failed to unlock element: {error}")

    async def move_cursor(self, page: int, x: float, y: float):
        """
        Broadcast cursor position.

        Args:
            page: Current page number.
            x: X coordinate.
            y: Y coordinate.
        """
        await self.sio.emit(
            "cursor:move",
            {"page": page, "x": x, "y": y},
        )

        logger.debug(f"Cursor moved to page {page}, position ({x}, {y})")

    async def broadcast_update(
        self,
        update_type: str,
        affected_elements: list[str],
        affected_pages: list[int],
        data: dict,
    ):
        """
        Broadcast a document update.

        Args:
            update_type: Type of update (e.g., 'element_modified').
            affected_elements: List of affected element UUIDs.
            affected_pages: List of affected page numbers.
            data: Additional update data.
        """
        await self.sio.emit(
            "document:update",
            {
                "update_type": update_type,
                "affected_elements": affected_elements,
                "affected_pages": affected_pages,
                "timestamp": datetime.utcnow().isoformat(),
                "data": data,
            },
        )

        logger.info(
            f"Broadcast update: {update_type} "
            f"(elements: {len(affected_elements)}, pages: {affected_pages})"
        )

    async def disconnect(self):
        """Disconnect from the server."""
        logger.info("Disconnecting...")
        await self.sio.disconnect()

    async def wait(self):
        """Wait for the client to disconnect."""
        await self.sio.wait()


async def main():
    """Main entry point for the example client."""
    parser = argparse.ArgumentParser(
        description="Giga-PDF WebSocket Collaboration Client"
    )
    parser.add_argument(
        "--server-url",
        default="http://localhost:8000/ws",
        help="WebSocket server URL (default: http://localhost:8000/ws)",
    )
    parser.add_argument(
        "--document-id",
        required=True,
        help="Document UUID to collaborate on",
    )
    parser.add_argument(
        "--token",
        required=True,
        help="JWT authentication token",
    )
    parser.add_argument(
        "--demo",
        action="store_true",
        help="Run demo interactions (lock, move cursor, etc.)",
    )

    args = parser.parse_args()

    # Create client
    client = CollaborationClient(
        server_url=args.server_url,
        document_id=args.document_id,
        auth_token=args.token,
    )

    try:
        # Connect to server
        await client.connect()

        if args.demo:
            # Run demo interactions
            logger.info("\n=== Running demo interactions ===\n")

            # Move cursor around
            logger.info("Moving cursor...")
            for i in range(5):
                await client.move_cursor(1, 100 + i * 10, 200 + i * 10)
                await asyncio.sleep(1)

            # Try to lock an element
            logger.info("\nTrying to lock element...")
            test_element_id = "test-element-uuid"
            locked = await client.lock_element(test_element_id)

            if locked:
                # Broadcast an update
                logger.info("\nBroadcasting update...")
                await client.broadcast_update(
                    update_type="element_modified",
                    affected_elements=[test_element_id],
                    affected_pages=[1],
                    data={
                        "element_id": test_element_id,
                        "property": "text",
                        "value": "Updated text from demo",
                    },
                )

                await asyncio.sleep(2)

                # Unlock the element
                await client.unlock_element(test_element_id)

            logger.info("\n=== Demo complete ===\n")
            logger.info("Press Ctrl+C to exit")

        # Wait for disconnect
        await client.wait()

    except KeyboardInterrupt:
        logger.info("\nReceived interrupt, cleaning up...")
    except Exception as e:
        logger.error(f"Error: {e}", exc_info=True)
    finally:
        await client.disconnect()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/websocket_client.js
/**
 * Example WebSocket collaboration client for Giga-PDF (JavaScript/Node.js).
 *
 * This example demonstrates how to connect to the Giga-PDF collaboration
 * server and interact with other users in real-time.
 *
 * Installation:
 *   npm install socket.io-client
 *
 * Usage:
 *   node examples/websocket_client.js --document-id <uuid> --token <jwt>
 */

const io = require('socket.io-client');

/**
 * WebSocket collaboration client for Giga-PDF.
 */
class CollaborationClient {
  /**
   * Initialize collaboration client.
   *
   * @param {string} serverUrl - WebSocket server URL (e.g., http://localhost:8000/ws)
   * @param {string} documentId - Document UUID to collaborate on
   * @param {string} authToken - JWT authentication token
   */
  constructor(serverUrl, documentId, authToken) {
    this.serverUrl = serverUrl;
    this.documentId = documentId;
    this.authToken = authToken;
    this.sessionId = null;
    this.userColor = null;
    this.activeUsers = new Map();
    this.lockedElements = new Set();

    this.socket = io(serverUrl, {
      auth: {
        token: authToken,
      },
      transports: ['websocket', 'polling'],
    });

    this.setupEventHandlers();
  }

  /**
   * Set up Socket.IO event handlers.
   */
  setupEventHandlers() {
    // Connection events
    this.socket.on('connect', () => {
      console.log('Connected to collaboration server');
      this.joinDocument();
    });

    this.socket.on('disconnect', () => {
      console.log('Disconnected from collaboration server');
      this.activeUsers.clear();
      this.lockedElements.clear();
    });

    this.socket.on('connect_error', (error) => {
      console.error('Connection error:', error.message);
    });

    // User presence events
    this.socket.on('user:joined', (data) => {
      const { user_id, user_name, user_color } = data;

      this.activeUsers.set(user_id, {
        name: user_name,
        color: user_color,
      });

      console.log(`User joined: ${user_name} (${user_id}) with color ${user_color}`);
    });

    this.socket.on('user:left', (data) => {
      const { user_id, user_name } = data;

      this.activeUsers.delete(user_id);

      console.log(`User left: ${user_name} (${user_id})`);
    });

    // Element locking events
    this.socket.on('element:locked', (data) => {
      const { element_id, locked_by_user_name, expires_at } = data;

      this.lockedElements.add(element_id);

      console.log(`Element ${element_id} locked by ${locked_by_user_name} until ${expires_at}`);
    });

    this.socket.on('element:unlocked', (data) => {
      const { element_id } = data;

      this.lockedElements.delete(element_id);

      console.log(`Element ${element_id} unlocked`);
    });

    // Cursor tracking
    this.socket.on('cursor:moved', (data) => {
      const { user_id, user_name, page, x, y } = data;

      console.debug(`Cursor from ${user_name}: page ${page}, position (${x.toFixed(1)}, ${y.toFixed(1)})`);
    });

    // Document updates
    this.socket.on('document:updated', (data) => {
      const { user_name, update_type, affected_elements = [], affected_pages = [] } = data;

      console.log(
        `Document updated by ${user_name}: ${update_type} ` +
        `(elements: ${affected_elements.length}, pages: ${affected_pages})`
      );
    });
  }

  /**
   * Join the document collaboration room.
   */
  async joinDocument() {
    console.log(`Joining document ${this.documentId}...`);

    return new Promise((resolve, reject) => {
      this.socket.emit('join_document', { document_id: this.documentId }, (response) => {
        if (response.success) {
          this.sessionId = response.session.session_id;
          this.userColor = response.session.user_color;

          console.log('Joined document successfully!');
          console.log(`Session ID: ${this.sessionId}`);
          console.log(`Your color: ${this.userColor}`);

          // Log active users
          const activeUsers = response.active_users || [];
          if (activeUsers.length > 0) {
            console.log(`Active users (${activeUsers.length}):`);
            activeUsers.forEach((user) => {
              console.log(`  - ${user.user_name} (${user.user_id}) with color ${user.user_color}`);
              this.activeUsers.set(user.user_id, {
                name: user.user_name,
                color: user.user_color,
              });
            });
          } else {
            console.log('You are the first user in this document');
          }

          // Log active locks
          const activeLocks = response.active_locks || [];
          if (activeLocks.length > 0) {
            console.log(`Active locks (${activeLocks.length}):`);
            activeLocks.forEach((lock) => {
              console.log(`  - Element ${lock.element_id} locked by ${lock.locked_by_user_id}`);
              this.lockedElements.add(lock.element_id);
            });
          }

          resolve(response);
        } else {
          const error = response.error || 'Unknown error';
          console.error(`Failed to join document: ${error}`);
          reject(new Error(error));
        }
      });
    });
  }

  /**
   * Leave the document collaboration room.
   */
  async leaveDocument() {
    console.log('Leaving document...');

    return new Promise((resolve, reject) => {
      this.socket.emit('leave_document', {}, (response) => {
        if (response.success) {
          console.log('Left document successfully');
          this.activeUsers.clear();
          this.lockedElements.clear();
          resolve(response);
        } else {
          const error = response.error || 'Unknown error';
          console.error(`Failed to leave document: ${error}`);
          reject(new Error(error));
        }
      });
    });
  }

  /**
   * Lock an element for editing.
   *
   * @param {string} elementId - Element UUID to lock
   * @returns {Promise<boolean>} True if lock acquired
   */
  async lockElement(elementId) {
    console.log(`Locking element ${elementId}...`);

    return new Promise((resolve) => {
      this.socket.emit('element:lock', { element_id: elementId }, (response) => {
        if (response.success) {
          console.log(`Element locked until ${response.lock.expires_at}`);
          this.lockedElements.add(elementId);
          resolve(true);
        } else {
          const error = response.error || 'Unknown error';
          const lockedBy = response.locked_by;
          const expiresAt = response.expires_at;

          console.warn(
            `Failed to lock element: ${error} ` +
            `(locked by ${lockedBy} until ${expiresAt})`
          );
          resolve(false);
        }
      });
    });
  }

  /**
   * Unlock an element.
   *
   * @param {string} elementId - Element UUID to unlock
   */
  async unlockElement(elementId) {
    console.log(`Unlocking element ${elementId}...`);

    return new Promise((resolve, reject) => {
      this.socket.emit('element:unlock', { element_id: elementId }, (response) => {
        if (response.success) {
          console.log('Element unlocked successfully');
          this.lockedElements.delete(elementId);
          resolve(response);
        } else {
          const error = response.error || 'Unknown error';
          console.warn(`Failed to unlock element: ${error}`);
          reject(new Error(error));
        }
      });
    });
  }

  /**
   * Broadcast cursor position.
   *
   * @param {number} page - Current page number
   * @param {number} x - X coordinate
   * @param {number} y - Y coordinate
   */
  moveCursor(page, x, y) {
    this.socket.emit('cursor:move', { page, x, y });
    console.debug(`Cursor moved to page ${page}, position (${x}, ${y})`);
  }

  /**
   * Broadcast a document update.
   *
   * @param {string} updateType - Type of update
   * @param {string[]} affectedElements - List of affected element UUIDs
   * @param {number[]} affectedPages - List of affected page numbers
   * @param {object} data - Additional update data
   */
  broadcastUpdate(updateType, affectedElements, affectedPages, data) {
    this.socket.emit('document:update', {
      update_type: updateType,
      affected_elements: affectedElements,
      affected_pages: affectedPages,
      timestamp: new Date().toISOString(),
      data: data,
    });

    console.log(
      `Broadcast update: ${updateType} ` +
      `(elements: ${affectedElements.length}, pages: ${affectedPages})`
    );
  }

  /**
   * Disconnect from the server.
   */
  disconnect() {
    console.log('Disconnecting...');
    this.socket.disconnect();
  }
}

/**
 * Sleep utility function.
 *
 * @param {number} ms - Milliseconds to sleep
 * @returns {Promise<void>}
 */
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Main entry point for the example client.
 */
async function main() {
  // Parse command line arguments
  const args = process.argv.slice(2);
  const serverUrl = getArg(args, '--server-url', 'http://localhost:8000/ws');
  const documentId = getArg(args, '--document-id');
  const token = getArg(args, '--token');
  const runDemo = args.includes('--demo');

  if (!documentId) {
    console.error('Error: --document-id is required');
    console.log('Usage: node websocket_client.js --document-id <uuid> --token <jwt> [--demo]');
    process.exit(1);
  }

  if (!token) {
    console.error('Error: --token is required');
    console.log('Usage: node websocket_client.js --document-id <uuid> --token <jwt> [--demo]');
    process.exit(1);
  }

  // Create client
  const client = new CollaborationClient(serverUrl, documentId, token);

  // Handle graceful shutdown
  process.on('SIGINT', async () => {
    console.log('\nReceived interrupt, cleaning up...');
    client.disconnect();
    process.exit(0);
  });

  // Wait for connection
  await sleep(2000);

  if (runDemo) {
    console.log('\n=== Running demo interactions ===\n');

    // Move cursor around
    console.log('Moving cursor...');
    for (let i = 0; i < 5; i++) {
      client.moveCursor(1, 100 + i * 10, 200 + i * 10);
      await sleep(1000);
    }

    // Try to lock an element
    console.log('\nTrying to lock element...');
    const testElementId = 'test-element-uuid';
    const locked = await client.lockElement(testElementId);

    if (locked) {
      // Broadcast an update
      console.log('\nBroadcasting update...');
      client.broadcastUpdate(
        'element_modified',
        [testElementId],
        [1],
        {
          element_id: testElementId,
          property: 'text',
          value: 'Updated text from demo',
        }
      );

      await sleep(2000);

      // Unlock the element
      await client.unlockElement(testElementId);
    }

    console.log('\n=== Demo complete ===\n');
    console.log('Press Ctrl+C to exit');
  }
}

/**
 * Get command line argument value.
 *
 * @param {string[]} args - Command line arguments
 * @param {string} name - Argument name
 * @param {string} defaultValue - Default value if not found
 * @returns {string|undefined}
 */
function getArg(args, name, defaultValue = undefined) {
  const index = args.indexOf(name);
  if (index !== -1 && index + 1 < args.length) {
    return args[index + 1];
  }
  return defaultValue;
}

// Run main function
if (require.main === module) {
  main().catch((error) => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

module.exports = { CollaborationClient };
